README


This archive contains the distribution Text-Keywords,
version 0.004:

  Setup Text::Keywords::Container for analyzing some text

This software is copyright (c) 2011 by L<Raudssus Social Software|http://www.raudssus.de/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Text-Keywords

0.004     2011-02-03 05:35:24 Europe/Berlin
	- Fixed regexp for matching (again..)

0.003     2011-01-22 04:26:37 Europe/Berlin
	- Changed regexp again to be more consistent

0.002     2011-01-22 04:02:32 Europe/Berlin
	- Added params to container
	- Fixed regexp for matching

0.001     2011-01-10 06:12:42 Europe/Berlin
	- Initial official release


lib/Text/Keywords.pm
package Text::Keywords;
BEGIN {
  $Text::Keywords::AUTHORITY = 'cpan:GETTY';
}
BEGIN {
  $Text::Keywords::VERSION = '0.004';
}
# ABSTRACT: Setup Text::Keywords::Container for analyzing some text

use Moo;

has containers => (
	is => 'rw',
	default => sub {[]},
);

sub from {
	my ( $self, $primary, $secondary ) = @_;
	
	my @founds;

	for (@{$self->containers}) {
		push @founds, $_->find_keywords($primary, $secondary);
	}
	
	@founds = $self->modify_founds(@founds);

	return @founds;
}

sub modify_founds {
	my ( $self, @founds ) = @_;
	return @founds;
}

1;

__END__
=pod

=head1 NAME

Text::Keywords - Setup Text::Keywords::Container for analyzing some text

=head1 VERSION

version 0.004

=head1 AUTHOR

Torsten Raudssus <torsten@raudssus.de> L<http://www.raudssus.de/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by L<Raudssus Social Software|http://www.raudssus.de/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Text/Keywords/Found.pm
package Text::Keywords::Found;
BEGIN {
  $Text::Keywords::Found::AUTHORITY = 'cpan:GETTY';
}
BEGIN {
  $Text::Keywords::Found::VERSION = '0.004';
}
# ABSTRACT: Class for a keyword found over a specific Text::Keywords::Container

use Moo;

has keyword => (
	is => 'ro',
	required => 1,
);

has found => (
	is => 'ro',
	required => 1,
);

has matches => (
	is => 'ro',
	default => sub {[]},
);

has in_primary => (
	is => 'ro',
	required => 1,
);

has in_secondary => (
	is => 'ro',
	required => 1,
);

has container => (
	is => 'ro',
	required => 1,
);

has list => (
	is => 'ro',
	required => 1,
);

1;
__END__
=pod

=head1 NAME

Text::Keywords::Found - Class for a keyword found over a specific Text::Keywords::Container

=head1 VERSION

version 0.004

=head1 AUTHOR

Torsten Raudssus <torsten@raudssus.de> L<http://www.raudssus.de/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by L<Raudssus Social Software|http://www.raudssus.de/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Text/Keywords/Container.pm
package Text::Keywords::Container;
BEGIN {
  $Text::Keywords::Container::AUTHORITY = 'cpan:GETTY';
}
BEGIN {
  $Text::Keywords::Container::VERSION = '0.004';
}
# ABSTRACT: Class for a container of serveral Text::Keywords::List

use 5.010;
use Moo;
use Text::Keywords::Found;

has lists => (
	is => 'ro',
	required => 1,
);

has shuffle => (
	is => 'ro',
	default => sub { 1 },
);

has use_secondary => (
	is => 'ro',
	default => sub { 1 },
);

has params => (
	is => 'ro',
	default => sub {{}},
);

sub find_keywords {
	my ( $self, $primary, $secondary ) = @_;
	$primary = $secondary if !$primary;
	return () if !$primary;
	my @founds;
	my @keywordlists;
	my $klpos = 0;
	for (@{$self->lists}) {
		push @keywordlists, [ $_, $_->count ];
	}
	while (@keywordlists) {
		my $klcount = scalar @keywordlists;
		my $idx = $klpos % $klcount;
		my $kl = $keywordlists[$idx];
		$klpos++;
		for my $keyword (@{$kl->[0]->keywords}[$kl->[0]->count - $kl->[1],$kl->[0]->count - 1]) {
			$kl->[1]--;
			my $found;
			splice(@keywordlists, $idx, 1) if (!$kl->[1]);
			my $rx = qr/(^|[^\w]|_)($keyword)/i;
			my @text_found;
			my $primary_str = $primary;
			push (@text_found, [$primary_str =~ $rx]), $primary_str =~ s{$rx}{} while $primary_str =~ $rx;
			my @secondary_found;
			if ($secondary && $self->use_secondary) {
				my $con_str = $secondary;
				push (@secondary_found, [$con_str =~ $rx]), $con_str =~ s{$rx}{} while $con_str =~ $rx;
			}
			for (@text_found) {
				$found = 1;
				my @matches = @{$_};
				shift @matches;
				my $found_string = shift @matches;
				my $already_found = 0;
				for my $cur_found (@founds) {
					if ($cur_found->found eq $found_string and @{$cur_found->matches} ~~ @matches) {
						$already_found = 1; last;
					}
				}
				next if $already_found;
				my $found_in_secondary = 0;
				my $cidx = 0;
				for (@secondary_found) {
					my @cmatches = @{$_};
					shift @cmatches;
					my $found_cstring = shift @cmatches;
					if ($found_string eq $found_cstring and @matches ~~ @cmatches) {
						delete $secondary_found[$cidx];
					}
					$cidx++;
				}
				push @founds, $self->_new_found($found_string,1,$found_in_secondary,$kl->[0],$keyword,\@matches);
			}
			my @still_secondary_found;
			if (@text_found) {
				for (@secondary_found) {
					push @still_secondary_found, $_ if ($_);
				}
			} else {
				@still_secondary_found = @secondary_found;
			}
			for (@still_secondary_found) {
				$found = 1;
				my @matches = @{$_};
				shift @matches;
				my $found_string = shift @matches;
				my $already_found = 0;
				for my $cur_found (@founds) {
					if ($cur_found->found eq $found_string and @{$cur_found->matches} ~~ @matches) {
						$already_found = 1; last;
					}
				}
				next if $already_found;
				push @founds, $self->_new_found($found_string,0,1,$kl->[0],$keyword,\@matches);
			}
			last if $self->shuffle and $found;
		}
	}
	return @founds;
}

sub _new_found {
	my ( $self, $found, $in_primary, $in_secondary, $keywordlist, $keyword, $matches ) = @_;
	return Text::Keywords::Found->new({
		keyword => $keyword,
		found => $found,
		list => $keywordlist,
		matches => $matches,
		in_primary => $in_primary ? 1 : 0,
		in_secondary => $in_secondary ? 1 : 0,
		container => $self
	});
}

1;
__END__
=pod

=head1 NAME

Text::Keywords::Container - Class for a container of serveral Text::Keywords::List

=head1 VERSION

version 0.004

=head1 AUTHOR

Torsten Raudssus <torsten@raudssus.de> L<http://www.raudssus.de/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by L<Raudssus Social Software|http://www.raudssus.de/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

BEGIN { require 5.010; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Setup Text::Keywords::Container for analyzing some text',
  'AUTHOR' => 'Torsten Raudssus <torsten@raudssus.de> L<http://www.raudssus.de/>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0.96'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Text-Keywords',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Text::Keywords',
  'PREREQ_PM' => {
    'Moo' => '0.009005'
  },
  'VERSION' => '0.004',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



